### PR TITLE
MAINT: port numpy#13188 for np_datetime simplification

### LINKED ISSUE #26500: MAINT: port numpy#13188 for np_datetime simplification
our _libs/tslibs/src/np_datetime.c is based off of the numpy version.  A recent PR there made a pretty nice simplfication to the relevant code.  We should do the same.

https://github.com/numpy/numpy/pull/13188/files

### PR DESCRIPTION
**Pull request by xcz011**
Bring numpy changes about emulating the behavior of python's divmod to
pandas.

- [x] closes #26500
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by gfyoung — APPROVED** [review]
Nice!

@jbrockmendel : If you want to have a look before merging

File: pandas/_libs/tslibs/src/datetime/np_datetime.c [hunk 1]
```diff
@@ -498,6 +498,25 @@ npy_datetime npy_datetimestruct_to_datetime(NPY_DATETIMEUNIT base,
     return ret;
 }
 
+/*
+ * Computes the python `ret, d = divmod(d, unit)`.
+ *
+ * Note that GCC is smart enough at -O2 to eliminate the `if(*d < 0)` branch
+ * for subsequent calls to this command - it is able to deduce that `*d >= 0`.
+ */
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/src/datetime/np_datetime.c]: Can you add to this comment pointing back to the numpy PR.

**Comment by jbrockmendel** [comment]
Generally looks good.  Were there any parts that weren't copy/paste?

File: pandas/_libs/tslibs/src/datetime/np_datetime.c (hunk 1 above)
Comment (xcz011) [reply pandas/_libs/tslibs/src/datetime/np_datetime.c]: just added it. Thanks! 😄

**Comment by xcz011** [comment]
> Generally looks good. Were there any parts that weren't copy/paste?

All changes are copied from numpy PR. I changed a few line style since those lines over 80 characters.

**Comment by jreback** [comment]
lgtm. @jbrockmendel merge when ready.

**Comment by jbrockmendel** [comment]
Thanks @xcz011

**Pull request merged**